pandas/core/frame.py

```diff
@@ -6804,54 +6810,58 @@ def cov(self, min_periods=None):
 
         return self._constructor(baseCov, index=idx, columns=cols)
 
-    def corrwith(self, other, axis=0, drop=False):
+    def corrwith(self, other, axis=0, drop=False, method='pearson'):
         """
-        Compute pairwise correlation between rows or columns of two DataFrame
-        objects.
+        Compute pairwise correlation between rows or columns of DataFrame
+        with rows or columns of Series or DataFrame.  DataFrames are first
+        aligned along both axes before computing the correlations.
 
         Parameters
         ----------
         other : DataFrame, Series
         axis : {0 or 'index', 1 or 'columns'}, default 0
             0 or 'index' to compute column-wise, 1 or 'columns' for row-wise
         drop : boolean, default False
-            Drop missing indices from result, default returns union of all
+            Drop missing indices from result
+        method : {'pearson', 'kendall', 'spearman'}
+            * pearson : standard correlation coefficient
+            * kendall : Kendall Tau correlation coefficient
+            * spearman : Spearman rank correlation
+
+            .. versionadded:: 0.24.0
 
         Returns
         -------
         correls : Series
+
+        See Also
+        -------
+        DataFrame.corr
         """
+        if method not in ['pearson', 'spearman', 'kendall']:
+            raise ValueError("method must be either 'pearson', "
+                             "'spearman', or 'kendall', '{method}' "
+                             "was supplied".format(method=method))
+
         axis = self._get_axis_number(axis)
         this = self._get_numeric_data()
 
         if isinstance(other, Series):
-            return this.apply(other.corr, axis=axis)
+            return this.apply(lambda x: other.corr(x, method=method),
+                              axis=axis)
 
         other = other._get_numeric_data()
-
         left, right = this.align(other, join='inner', copy=False)
 
-        # mask missing values
-        left = left + right * 0
-        right = right + left * 0
-
         if axis == 1:
             left = left.T
             right = right.T
 
-        # demeaned data
-        ldem = left - left.mean()
-        rdem = right - right.mean()
-
-        num = (ldem * rdem).sum()
-        dom = (left.count() - 1) * left.std() * right.std()
+        correl = (left.apply(lambda x:
+                             x.corr(right[x.name], method=method)))
```

Comment:
Does this break when left or right have duplicate columns? Can you add a test ensuring that it works? It seems like we support that on master.

Reply:
I'll test that out.  What would you say the expected behavior should be in this case?  Would it make sense to first drop duplicate columns from `self` and `other`?

Reply:
Yes, I believe this breaks with duplicate columns.  There might be a way to still rely on `apply`, but if not I'll try to think of something else.  In the worst case a `for` loop should work, but I'd like to avoid that if possible.

Reply:
@TomAugspurger I think this is actually an error on `master`, but I may have a fix.

```
import numpy as np
import pandas as pd

df1 = pd.DataFrame(np.random.normal(size=(10, 3)))
df2 = pd.DataFrame(np.random.normal(size=(10, 3)))
df2 = pd.concat((df2, pd.DataFrame(np.random.randn(10))), axis=1)

df1.corrwith(df2)
```

A possible fix within `corrwith` is to create `correl` by applying a `map` to the `zipped` columns

```
pd.Series(map(lambda x: pd.Series(x[0]).corr(pd.Series(x[1]), method="spearman"),
              zip(left.values.T, right.values.T)),
          index=left.columns)
```

I can commit this as a change if you think it makes sense.